Improvement/2016 brand the dex gui to meet the design

ChengYanJin opened the pull request:
**Component**: Dex, UI

**Context**: 
As I user, I want to have the same user experience as Metalk8s UI when I log in.

**Summary**:
We create a configMap and mount it as volume in the web/theme/scality
```
#!jinja | metalk8s_kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: dex-login
  namespace: metalk8s-auth
data:
  styles.css: @@styles
binaryData:
  favicon.png: @@favicon
  logo.png: @@branding
```
**Acceptance criteria**: 
1. ./doit.sh vagrant_up
2. Get the control plane ip
3.  Go to the login page
```https://<control_plane_ip>:8443/oidc/auth?audience=&amp;client_id=metalk8s-ui&amp;redirect_uri=https%3A%2F%2F<control_plane_ip>%3A8443%2Foauth2%2Fcallback&amp;response_type=id_token&amp;scope=openid+profile+email+offline_access&amp;nonce=abcde```
![image](https://user-images.githubusercontent.com/18453133/69349785-add13800-0c78-11ea-830a-24f042bbbf19.png)

Closes: #2016

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet commented:
/after_pull_request=2063

bert-e commented:
# Waiting for other pull request(s)



The current pull request is locked by the **after_pull_request** option.

In order for me to merge this pull request, run the following actions first:


:arrow_right: Merge the `OPEN` pull request:

* pull request #2063





Alternatively, delete all the **after_pull_request** comments from this pull request.





*The following options are set:* **after_pull_request**

gdemonet commented:
/after_pull_request=2025

bert-e commented:
# Waiting for other pull request(s)



The current pull request is locked by the **after_pull_request** option.

In order for me to merge this pull request, run the following actions first:


:arrow_right: Merge the `OPEN` pull requests:

* pull request #2025

* pull request #2063





Alternatively, delete all the **after_pull_request** comments from this pull request.





*The following options are set:* **after_pull_request**

gdemonet commented:
@ChengYanJin you'll need to rebase on `development/2.5` once #2063 gets merged (and probably will need the code from #2025 as well)

bert-e commented:
# Waiting for other pull request(s)



The current pull request is locked by the **after_pull_request** option.

In order for me to merge this pull request, run the following actions first:


:arrow_right: Merge the `OPEN` pull request:

* pull request #2025





Alternatively, delete all the **after_pull_request** comments from this pull request.





*The following options are set:* **after_pull_request**

ChengYanJin committed:
ui/login: Add the UI resources to customize the Dex login page

The three files will be mounted in web/themes/scality in order to render our customized theme

Refs: #2016

ChengYanJin committed:
buildchain/salt: Create the configMap dex-login including the UI resources

- change the buildchain in order to get the content of three UI resource in ConfigMap
- add two function salt_embed_bytes and salt_embed_text to get the good format yaml

Refs: #2016

ChengYanJin committed:
salt/dex: Add and mount the Dex branding volumes

Generate chart.sls from dex.yaml

Refs: #2016

slaperche-scality reviewed (commented):
LGTM for the buildchain part.
I tested it and it works.

Good job :+1:

gdemonet commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -153,6 +153,25 @@ def _get_parts(self) -> Iterator[str]:
         return parts
 
 
+def salt_embed_text(path: Path, indent: int) -> str:
+    """Return the content a text file in usable form by Salt."""
+    data = ['|']
+    with path.open(encoding='utf-8') as fp:
+        for line in fp:
+            data.append('{}{}'.format(' '*indent, line.rstrip()))
+    return '\n'.join(data)
+
+
+def salt_embed_bytes(path: Path, indent: int) -> str:
+    """Return the content of a binary file in usable form by Salt."""
+    data = ['|']
+    bytestring = path.read_bytes()
+    b64data = base64.encodebytes(bytestring).decode('utf-8')
+    for line in b64data.split('\n'):
+        data.append('{}{}'.format(' '*indent, line.rstrip()))
+    return '\n'.join(data)
```
Mmpf, that's a strange way to do things. Both could rely on `yaml.dump` (see https://github.com/scality/metalk8s/pull/2042/commits/1ada06c5769e987afc7473cb58250419e413b5bd for an example of how to customize representers). But let's keep it for a debt ticket.

gdemonet commented on charts/dex.yaml:
```diff
@@ -52,8 +52,8 @@ config:
   frontend:
-    theme: "coreos" #metalk8s-ui
-    # dir: /web/themes/custom/
+    theme: "scality"
+    issuer: "MetalK8s"
```
Must have took you quite some time to find out how this works, as it's not documented at all... I had to read the Go sources to understand what it does... Maybe worth a PR upstream to clarify this usage?

gdemonet commented on salt/metalk8s/addons/dex/deployed/login.sls.in:
```diff
@@ -0,0 +1,14 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: dex-login
```
Let's call this file `theme-configmap.sls` for clarity, or just `configmap` if you prefer. Login is too vague IMHO.

gdemonet commented on salt/metalk8s/addons/dex/deployed/login.sls.in:
```diff
@@ -0,0 +1,14 @@
+#!jinja | metalk8s_kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes
```
After #2050 we'll stop using `kubeconfig` and `context` completely, so I don't know how you want to handle that, maybe rebase and remove it right now...

TeddyAndrieux replied:
We still support it just it's not needed so we can merge like this and remove it after

gdemonet reviewed (changes requested):
Minor changes, good job! We'll need to think about a different approach though (I guess fetch the assets from MetalUI directly could work, so that if the theme changes, this one will as well)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet







*The following options are set:* **after_pull_request**

ChengYanJin commented:
> Minor changes, good job! We'll need to think about a different approach though (I guess fetch the assets from MetalUI directly could work, so that if the theme changes, this one will as well)

Hi Guillaume, thanks for the review!

Actually I fetch different asset for Dexlogin and Metalk8sUI. e.g For scality logo, we use branding.svg => MetalK8s and logo.png => Dexlogin. 
The reason is that I need to follow the default file name of Dex.

I moved the dexlogin assets in the ui/public/brand/assets/login. Wondering does it make sense to you?

gdemonet commented:
> Actually I fetch different asset for Dexlogin and Metalk8sUI. e.g For scality logo, we use branding.svg => MetalK8s and logo.png => Dexlogin. 
> The reason is that I need to follow the default file name of Dex.

Oh OK. Actually my point was that, instead of mounting those assets in the Dex container, we could find some way to let them be referenced by URL instead (such URL pointing to Metal UI). But really not urgent. I think I've seen a `logoURL` configuration option in some examples, for instance.

> I moved the dexlogin assets in the ui/public/brand/assets/login. Wondering does it make sense to you?

Yeah it will make sense if we manage to serve them with Metal UI (and not store them in a ConfigMap at build time). So let's keep it, good idea :)

gdemonet commented:
> Actually I fetch different asset for Dexlogin and Metalk8sUI. e.g For scality logo, we use branding.svg => MetalK8s and logo.png => Dexlogin. 
> The reason is that I need to follow the default file name of Dex.

Oh OK. Actually my point was that, instead of mounting those assets in the Dex container, we could find some way to let them be referenced by URL instead (such URL pointing to Metal UI). But really not urgent. I think I've seen a `logoURL` configuration option in some examples, for instance.

> I moved the dexlogin assets in the ui/public/brand/assets/login. Wondering does it make sense to you?

Yeah it will make sense if we manage to serve them with Metal UI (and not store them in a ConfigMap at build time). So let's keep it, good idea :)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **after_pull_request**

ChengYanJin commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve, after_pull_request**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye chengyanjin.

NicolasT commented on charts/dex.yaml:
```diff
@@ -31,14 +31,14 @@ ingress:
+extraVolumes:
+  - name: dex-login
+    configMap:
+      name: dex-login
```
I wish we'd have given this a less generic name like `dex-scality-theme` or whatever.

ChengYanJin replied:
> I wish we'd have given this a less generic name like `dex-scality-theme` or whatever.

I will rename it in the 2nd iteration of authentication.

NicolasT commented:
Don't postpone renaming of things. It's a hassle because of extra
upgrade/downgrade automation required.

On Thu, 19 Dec 2019, 16:27 ChengYanJin, <notifications@github.com> wrote:

> *@ChengYanJin* commented on this pull request.
> ------------------------------
>
> In charts/dex.yaml
> <https://github.com/scality/metalk8s/pull/2062#discussion_r359918296>:
>
> > @@ -31,14 +31,14 @@ ingress:
>    hosts:
>      - null
>
> -# extraVolumes:
> -#   - name: theme
> -#     configMap:
> -#        name: dex-branding
> +extraVolumes:
> +  - name: dex-login
> +    configMap:
> +      name: dex-login
>
> I wish we'd have given this a less generic name like dex-scality-theme or
> whatever.
>
> I will rename it in the 2nd iteration of authentication.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scality/metalk8s/pull/2062?email_source=notifications&email_token=AAAZBTO3QWYWNRXJJBYRZHLQZOHGPA5CNFSM4JQDV2TKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPZUBKQ#discussion_r359918296>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAZBTKKPSDEEPXTHK5U7FTQZOHGPANCNFSM4JQDV2TA>
> .
>
